### PR TITLE
remote dropship computers are no longer hard coded to dropship id

### PR DESCRIPTION
**Pull request by mullenpaul**
# About the pull request

Remote dropship flight control terminals are no longer hard coded. The remote controls in CIC have been removed from one per dropship to one console. The dropship to control can be selected in TGUI.

Groundside remote terminals will now randomly select a marine dropship to call down.

# Explain why it's good for the game
By removing hardcoded references to the alamo and normandy, we open the ability for the marines to select alternative dropships.

# Testing Photographs and Procedure
<details>
<summary>Screenshots & Videos</summary>

Put screenshots and videos here with an empty line between the screenshots and the `<details>` tags.

</details>

# Changelog
:cl:
ui: added section in dropship flight computer for some terminals to select which dropship to control
maptweak: reworked CIC remote terminals
refactor: some remote terminals can control different dropships
balance: queen will randomly select a dropship to call down to the lz on hijack
balance: when dropship exists on primary landing pad, the queen can't call down another dropship
balance: when a dropship is enroute to the primary landing pad, the queen can't call another dropship
/:cl:

### CODE REVIEW

File: code/modules/shuttle/computers/dropship_computer.dm [hunk 1]
```diff
@@ -211,6 +225,11 @@
 
 
 /obj/structure/machinery/computer/shuttle/dropship/flight/attack_alien(mob/living/carbon/xenomorph/xeno)
+	// if the shuttleid is null or the shuttleid references a shuttle that has been removed from play, pick one
+	if(!shuttleId || !SSshuttle.hasShuttle(shuttleId))
```
Comment (Drulikar) [inline code/modules/shuttle/computers/dropship_computer.dm:229]: As far as I can tell, hijack can only occur for the primary LZ, and the primary LZ is already going to be assigned to the alamo; so hijack will still occur as it has.

I also tried sending the alamo to infinite transit and that just causes a different runtime: 
image

**Review by Drulikar — CHANGES_REQUESTED** [review]
All human interactions with shuttle computers now throw exceptions: 
image
(Tested both the dropship remote control console in command, and the dropship navigation computer at LZ1 on LV624)

Additionally should it pick say the pizza shuttle, it would be funny but uh stuffing an average sized hive in this will be something... 
image

I don't think I would mind this is its rare though, but in general it should be a decently sized shuttle.

**Comment by mullenpaul** [comment]
> I don't think I would mind this is its rare though, but in general it should be a decently sized shuttle.

The alternative shuttle selection is just instances of marine_dropship - this should not happen. It should only currently select the alamo or the normandy. We can adjust the logic in the future to include additional shuttles.

File: code/modules/shuttle/computers/dropship_computer.dm (hunk 1 above)
Comment (mullenpaul) [reply code/modules/shuttle/computers/dropship_computer.dm:229]: I removed the overrides in the lz1 and lz2 subclasses of the flight computer which would hard code it in. I'll work on the runtimes - thanks for the testing

File: maps/map_files/USS_Almayer/USS_Almayer.dmm
Comment (Nanu308) [inline maps/map_files/USS_Almayer/USS_Almayer.dmm:1]: mapping seems fine

**Review by Nanu308 — APPROVED** [review]

File: code/modules/shuttle/computers/dropship_computer.dm [hunk 2]
```diff
@@ -128,6 +130,18 @@
 	.["max_refuel_duration"] = shuttle.rechargeTime / 10
 	.["max_engine_start_duration"] = shuttle.ignitionTime / 10
 	.["door_data"] = list("port", "starboard", "aft")
+	.["can_change_shuttle"] = can_change_shuttle
+	if(can_change_shuttle)
+		.["alternative_shuttles"] = alternative_shuttles()
```
[the remark this answers is not in the document]
Comment (mullenpaul) [reply code/modules/shuttle/computers/dropship_computer.dm:135]: This makes sense

File: code/modules/shuttle/computers/dropship_computer.dm [hunk 3]
```diff
@@ -507,18 +529,38 @@
...
-			return
+			return TRUE
+		if ("change_shuttle")
+			var/list/alternatives = alternative_shuttles()
+			var/new_shuttle = params["new_shuttle"]
+			var/found = FALSE
+			for(var/alt_shuttle in alternatives)
+				if(alt_shuttle["id"] == new_shuttle)
+					found = TRUE
+			if(found)
```
[the remark this answers is not in the document]
Comment (mullenpaul) [reply code/modules/shuttle/computers/dropship_computer.dm:557]: d'oh - good catch

File: tgui/packages/tgui/interfaces/DropshipFlightControl.tsx [hunk 4]
```diff
@@ -299,18 +307,15 @@ const PlayLaunchAnnouncementAlarm = () => {
       icon="rocket"
       onClick={() => {
         act('play_launch_announcement_alarm');
+        act('button-push');
```
[the remark this answers is not in the document]
Comment (mullenpaul) [reply tgui/packages/tgui/interfaces/DropshipFlightControl.tsx:310]: We can make the volume louder to make it more noticeable but I just wanted some sort of audible feedback for clicking a key

File: code/modules/shuttle/computers/dropship_computer.dm [hunk 5]
```diff
@@ -128,6 +130,18 @@
...
+/obj/structure/machinery/computer/shuttle/dropship/flight/proc/alternative_shuttles()
+	. = list()
+	for(var/obj/docking_port/mobile/marine_dropship/shuttle in SSshuttle.mobile)
+		. += list(
+			list(
+				"id" = shuttle.id, "name" = shuttle)
+		)
+
 
 /obj/structure/machinery/computer/shuttle/dropship/flight/attack_hand(mob/user)
```
[the remark this answers is not in the document]
Comment (mullenpaul) [reply code/modules/shuttle/computers/dropship_computer.dm:146]: I'll investigate this further

File: code/modules/shuttle/computers/dropship_computer.dm [hunk 6]
```diff
@@ -507,18 +529,38 @@
...
 			shuttle.playing_launch_announcement_alarm = TRUE
-			return
+			return TRUE
 		if ("stop_playing_launch_announcement_alarm")
+			if(!shuttle)
+				return FALSE
 			stop_playing_launch_announcement_alarm()
-			return
+			return TRUE
+		if ("change_shuttle")
```
[the remark this answers is not in the document]
Comment (mullenpaul) [reply code/modules/shuttle/computers/dropship_computer.dm:550]: I could make it go to a loading page whilst it loads the new static information. That would make it feel less broken

File: code/modules/shuttle/computers/dropship_computer.dm [hunk 7]
```diff
@@ -388,16 +408,19 @@
...
-		to_chat(user, SPAN_WARNING("The dropship isn't responding to controls."))
-		return
+	if (shuttle)
+		var/obj/structure/machinery/computer/shuttle/dropship/flight/comp = shuttle.getControlConsole()
+		if(comp.dropship_control_lost)
+			to_chat(user, SPAN_WARNING("The dropship isn't responding to controls."))
+			return
 
 	switch(action)
 		if("move")
```
[the remark this answers is not in the document]
Comment (mullenpaul) [reply code/modules/shuttle/computers/dropship_computer.dm:421]: Agreed. This is a remote function and should behave like it.

File: code/modules/shuttle/computers/dropship_computer.dm (hunk 5 above)
[the remark this answers is not in the document]
Comment (mullenpaul) [reply code/modules/shuttle/computers/dropship_computer.dm:148]: This has been resolved. I made a map change then didn't properly test it. Done now.

File: code/modules/shuttle/computers/dropship_computer.dm (hunk 7 above)
[the remark this answers is not in the document]
Comment (mullenpaul) [reply code/modules/shuttle/computers/dropship_computer.dm:430]: This should now be resolved

File: tgui/packages/tgui/interfaces/DropshipFlightControl.tsx [hunk 8]
```diff
@@ -14,7 +14,13 @@ interface AutomatedControl {
   ground_lz: null | string;
 }
 
+type ShuttleRef = {
+  name: string;
+  id: string;
+};
```
[the remark this answers is not in the document]
Comment (mullenpaul) [reply tgui/packages/tgui/interfaces/DropshipFlightControl.tsx:20]: They are fairly interchangeable in typescript. In this case I tend to use interface to describe the react props or state elements used and I normally use types to refer to sections of that. It's kinda arbitrary I admit. This should probably be an interface - source: https://blog.logrocket.com/types-vs-interfaces-typescript/

File: code/modules/shuttle/computers/dropship_computer.dm (hunk 6 above)
[the remark this answers is not in the document]
Comment (mullenpaul) [reply code/modules/shuttle/computers/dropship_computer.dm:559]: I'm not sure what we can really do about this - it takes a second to recalculate the static backend

[the remark this answers is not in the document]
Comment (SabreML) [reply code/modules/shuttle/computers/dropship_computer.dm:559]: As far as I can tell it's coming from the TGUI datum's two second `refresh_cooldown` not being finished, so maybe you could disable the selection buttons in the UI for around the same amount of time after they've been clicked?
https://github.com/cmss13-devs/cmss13/blob/8d8e164a4f6dbaf8f4e937e0b8d4e1b8e0521ce2/code/modules/tgui/tgui.dm#L201-L214
(called by `update_static_data()`)

[the remark this answers is not in the document]
Comment (Drulikar) [reply code/modules/shuttle/computers/dropship_computer.dm:559]: This act should probably do nothing if the current selection is already what we are selecting to cut back more on this issue.

**Review by Drulikar — CHANGES_REQUESTED** [review]
- Dropship computers for a dropship should be locked to the dropship they are:
image
In other words selecting a different dropship should be exclusive to `/obj/structure/machinery/computer/shuttle/dropship/flight/remote_control` not `/obj/structure/machinery/computer/shuttle/dropship/flight` or at the very least use the is_remote var (which in itself is very redundant)

Doing so seems to cause this runtime:
image

- Not sure if this is something specific I caused with what I did, and if its related to this PR, but I managed to get both shuttles to approach LZ1. Alamo was on autopilot, normandy I flew to hangar bay 1 then hangar bay 2. Queen interacting with LZ1 console refused to do anything until normandy was recharged which then called normandy while alamo was igniting for autopilot:
image
double dock

- Minor, but hangar bay computers probably ought to be generically named rather than their old purpose (its fine if they're preselected to be alamo/normandy): 
image

- Accessing a hijacked dropship via a remote console will then brick that console preventing you from selecting a working shuttle again: 
image
Closing the window and interacting with the computer then does nothing.

- This branch is particularly out of date and other shuttle code has been amended such as with 0c3c2058ddd678b8b3450f1dcf8c9a2582fcfdc1 where distress shuttles automatically launch. I recommend updating with master and make sure things are still working how you intend them to too. Of note I was able to as a CMB ERT, fly the shuttle to the admin z-level: 
image

**Comment by mullenpaul** [comment]
Changes I've made:

- Dropship selection not available on dropships.
- Changed name of remote computers in hangar to 'Remote dropship navigation computer'
- Remote consoles no longer bricked when selected dropship is hijacked
- Dropship selection disabled for 2 seconds after changing
- When the landing zone is occupied, the queen can't call the other dropship down to it
- When a dropship is enroute to a landing zone, the queen can't call a dropship down

**Comment by cm13-github** [comment]
This pull request has conflicts, please resolve those before we can evaluate the pull request.

**Comment by cm13-github** [comment]
Conflicts have been resolved. A maintainer will review the pull request shortly.

**Comment by cm13-github** [comment]
This pull request has conflicts, please resolve those before we can evaluate the pull request.

**Comment by cm13-github** [comment]
Conflicts have been resolved. A maintainer will review the pull request shortly.

**Comment by cm13-github** [comment]
This pull request has conflicts, please resolve those before we can evaluate the pull request.

**Comment by cm13-github** [comment]
Conflicts have been resolved. A maintainer will review the pull request shortly.

**Review by Drulikar — APPROVED** [review]
Code I think is fine, though needs testing to see if player catch anything else.

**Pull request merged**